Create a base Checkpointer and SyncCheckpointer and add dist barrier impl and

teja-rao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #156927
* __->__ #156926

In preparation to adding async checkpointing, this diff adds
1.  Change Checkpointer to an Abstract base class and adds a sync checkpointer implementation.
2. torch.distributed.barrier() as one of the barrier choices.

Differential Revision: [D77341314](https://our.internmc.facebook.com/intern/diff/D77341314/)

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

teja-rao committed:
Create a base Checkpointer and SyncCheckpointer and add dist barrier impl and 

In preparation to adding async checkpointing, this diff adds
1.  Change Checkpointer to an Abstract base class and adds a sync checkpointer implementation.
2. torch.distributed.barrier() as one of the barrier choices.

Differential Revision: [D77341314](https://our.internmc.facebook.com/intern/diff/D77341314/)

[ghstack-poisoned]

Skylion007 commented on torch/distributed/checkpoint/_experimental/checkpointer.py:
```diff
@@ -48,21 +123,24 @@ def save(
+        logger.debug("Saving checkpoint synchronously to %s", path)
         self._writer.write(state_dict, path, **kwargs)
+        return None
 
```
nit add typing.overload to prevent bugs from not overloading the methods during refactor.

Skylion007 commented on torch/distributed/checkpoint/_experimental/checkpointer.py:
```diff
@@ -12,20 +14,93 @@
+        *,
+        default_map_location: Any = None,
+        strict: bool = False,
+        **kwargs: Any,
```
```suggestion
        **kwargs: dict[str, Any],
```

teja-rao committed:
Update on "Create a base Checkpointer and SyncCheckpointer and add dist barrier impl and "


In preparation to adding async checkpointing, this diff adds
1.  Change Checkpointer to an Abstract base class and adds a sync checkpointer implementation.
2. torch.distributed.barrier() as one of the barrier choices.

Differential Revision: [D77341314](https://our.internmc.facebook.com/intern/diff/D77341314/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

teja-rao committed:
Update on "Create a base Checkpointer and SyncCheckpointer and add dist barrier impl and "


In preparation to adding async checkpointing, this diff adds
1.  Change Checkpointer to an Abstract base class and adds a sync checkpointer implementation.
2. torch.distributed.barrier() as one of the barrier choices.

Differential Revision: [D77341314](https://our.internmc.facebook.com/intern/diff/D77341314/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]
